helper/ldapifc/connection.go

```diff
@@ -0,0 +1,16 @@
+package ldapifc
+
+import (
+	"crypto/tls"
+	"github.com/go-ldap/ldap"
+)
+
+// Connection provides the functionality of an LDAP connection,
+// but through an interface.
+type Connection interface {
```

Comment:
Oh I thought that might bear some explanation! It actually is in use here: https://github.com/hashicorp/vault/pull/4260/files#diff-e03961031e1889b90ce8c3dfb1095ca1R37. 

In my original version of the LDAP client, I didn't use an interface and I used LDAP's statically available methods directly. However, when it came time to test the LDAP client, I didn't see any examples of tests around, and it looked like I could just use `httptest` for it because of stuff like this, for example: 
```
// DialTLS connects to the given address on the given network using tls.Dial
// and then returns a new Conn for the connection.
func DialTLS(network, addr string, config *tls.Config) (*Conn, error) {
	dc, err := net.DialTimeout(network, addr, DefaultTimeout)
	if err != nil {
		return nil, NewError(ErrorNetwork, err)
	}
	c := tls.Client(dc, config)
	err = c.Handshake()
	if err != nil {
		// Handshake error, close the established connection before we return an error
		dc.Close()
		return nil, NewError(ErrorNetwork, err)
	}
	conn := NewConn(c, true)
	conn.Start()
	return conn, nil
}
```
So, the best, easier-to-do options I could think of for testing were 1) use an interface that I can fulfill with a mock instead, or 2) just don't test?

I'm open to other approaches if you have ideas, that's just why I landed on this one.

Reply:
Thanks for linking those two articles, though! I actually hadn't seen them before, and I do have a tendency to do exactly what they say not to do, from a defensive coding perspective. However, now that I know it's not recommended, I can probably immediately revise some of the code like the `activedirectory.Client` to not return an interface, I'll look at that.